core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenTaskSource.java

```diff
@@ -119,180 +109,88 @@
-                    }
-                });
-            }
-            else {
-                splitLoaders.forEach(SplitLoader::start);
-            }
-        }
-        catch (Throwable t) {
-            try {
-                closer.close();
-            }
-            catch (Throwable closerFailure) {
-                if (closerFailure != t) {
-                    t.addSuppressed(closerFailure);
-                }
-            }
-            throw t;
+        checkState(!closed, "closed");
+
+        if (!initialized) {
+            initialize();
+            initialized = true;
         }
-    }
 
-    private SplitLoader createExchangeSplitLoader(PlanFragmentId fragmentId, PlanNodeId remoteSourceNodeId, ExchangeSplitSource splitSource)
-    {
-        return new SplitLoader(
-                splitSource,
-                executor,
-                ExchangeSplitSource::getSplitPartition,
-                new SplitLoader.Callback()
-                {
-                    @Override
-                    public void update(ListMultimap<Integer, Split> splits, boolean noMoreSplitsForFragment)
-                    {
-                        try {
-                            synchronized (assignerLock) {
-                                if (noMoreSplitsForFragment) {
-                                    finishedFragments.add(fragmentId);
-                                }
-                                boolean noMoreSplitsForRemoteSource = finishedFragments.containsAll(remoteSourceFragments.get(remoteSourceNodeId));
-                                assigner.assign(remoteSourceNodeId, splits, noMoreSplitsForRemoteSource).update(callback);
-                                if (noMoreSplitsForRemoteSource) {
-                                    finishedSources.add(remoteSourceNodeId);
-                                }
-                                if (finishedSources.containsAll(allSources)) {
-                                    assigner.finish().update(callback);
-                                }
-                            }
-                        }
-                        catch (Throwable t) {
-                            fail(t);
-                        }
-                    }
+        List<ListenableFuture<IdempotentSplitSource.SplitBatchReference>> futures = splitSources.stream()
+                .map(IdempotentSplitSource::getNext)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(toImmutableList());
 
-                    @Override
-                    public void failed(Throwable t)
-                    {
-                        fail(t);
-                    }
-                },
-                splitBatchSize,
-                getSplitTimeRecorder);
+        if (futures.isEmpty()) {
+            return immediateFuture(assigner.finish());
+        }
+
+        ListenableFuture<IdempotentSplitSource.SplitBatchReference> firstCompleted = whenAnyComplete(futures);
```

Comment:
This attaches a callback to each future on each call to `process`. If some sources are slow the list of callback can grow large.

You can avoid that by keeping `SettabeFuture<AssignmentResult` to be completed when new splits are discovered on the field. But I think it overall gets more complicated (you cannot easily exploit `IdempotentSplitSource` I think).
Maybe this is not a big deal.

Reply:
Yeah, that's a good point. Added a proxy future as an attempt to address the accumulating callbacks problem. Not sure if that's the cleanest solution though. Please take a look and let me know what you think.

Reply:
Tricky but should work. I really hope will I will not need to touch this code in the future :)